src/abi/anchor.rs

```diff
@@ -17,3 +29,343 @@ pub fn discriminator(namespace: &'static str, name: &str) -> Vec<u8> {
+                    docs: None,
+                    pda: None,
+                    relations: vec![],
+                })]
+            }
+        };
+
+        let mut args: Vec<IdlField> = Vec::with_capacity(func.params.len());
+        for (item_no, item) in func.params.iter().enumerate() {
+            let name = if item.id.is_none() || item.id.as_ref().unwrap().name.is_empty() {
+                format!("arg_{}", item_no)
+            } else {
+                item.id.as_ref().unwrap().name.clone()
+            };
+            args.push(IdlField {
+                name,
+                docs: None,
+                ty: type_manager.convert(&item.ty),
+            });
+        }
+
+        let returns = if func.returns.is_empty() {
+            None
+        } else if func.returns.len() == 1 {
+            Some(type_manager.convert(&func.returns[0].ty))
+        } else {
+            Some(type_manager.build_struct_for_return(func))
+        };
+
+        instructions.push(IdlInstruction {
+            name: if func.is_constructor() {
+                "new".to_string()
+            } else {
+                func.name.clone()
+            },
+            docs: tags,
+            accounts,
+            args,
+            returns,
+        });
+    }
+
+    instructions
+}
+
+/// This struct accounts all the user defined types used in the contract that need to be present
+/// in the IDL 'types' field.
+struct TypeManager<'a> {
+    namespace: &'a Namespace,
+    added_types: HashSet<Type>,
+    types: Vec<IdlTypeDefinition>,
+}
+
+impl TypeManager<'_> {
+    fn new(ns: &Namespace) -> TypeManager {
+        TypeManager {
+            namespace: ns,
+            added_types: HashSet::new(),
+            types: Vec::new(),
+        }
+    }
+
+    /// Retrieve all the custom types to fill the IDL struct
+    fn move_custom_types(&mut self) -> Vec<IdlTypeDefinition> {
+        let mut custom_types: Vec<IdlTypeDefinition> = Vec::new();
+        std::mem::swap(&mut custom_types, &mut self.types);
+        custom_types
+    }
+
+    /// Functions with multiple returns must return a struct in Anchor, so we build a return
+    /// struct containing all the returned types.
+    fn build_struct_for_return(&mut self, func: &Function) -> IdlType {
+        let mut fields: Vec<IdlField> = Vec::with_capacity(func.returns.len());
+        for (item_no, item) in func.returns.iter().enumerate() {
+            fields.push(IdlField {
+                name: format!("return_{}", item_no),
+                docs: None,
+                ty: self.convert(&item.ty),
+            });
+        }
+
+        let name = format!("{}_returns", func.name);
+        self.types.push(IdlTypeDefinition {
+            name: name.clone(),
+            docs: Some(vec![format!(
+                "Data structure to hold the multiple returns of function {}",
+                func.name
+            )]),
+            ty: IdlTypeDefinitionTy::Struct { fields },
+        });
```

Comment:
Are we checking for multiple definitions of the same name?

Reply:
We cannot overload structs in Solidity, can we?

Reply:
No. It is a good thing to do.

Reply:
Fixed